### ISSUE TITLE
Create option to get user by Vipps phone instead of email if already registered, thus consolidating users by Vipps ID

### ISSUE THREAD

**Issue by iverok** [vippsas/vipps-login-wordpress#14]
Requested here: https://wordpress.org/support/topic/login-with-vipps-creates-duplicate-users-when-user-changes-e-mail-address/

Especially relevant if Vipps is used as the primary login mechanism. Should be optional since some people (me) like having different accounts.

Easiest way would be to use a (filterable) function instead of get_user_by $email; with the default function using Vipps ID and Vipps Phone (with some canonicalization), and then the email. If the user found like this and the email doesn't match, the email should be changed.

However, if a different user exists with that email, that's going to be an issue; and any security issues with this (re-use of phone numbers?) need to be addressed.

**Comment by pdewouters** [comment]
Hi, have you made any progress on this? I saw. few updates since the ticket was opened, but I'm not sure whether they are related. 
I'm seeing this issue where WP users are created with their work email, and if they login with Vipps that uses a personal email. the user is not found

**Comment by iverok** [comment]
No, we haven't had the time. As mentioned above, the "normal" case is pretty simple, but error handling / conflict resolution is a bit worse and must be handled carefully.  The updates since this ticket are not related.

I'll have to add that of course the Vipps Phone number isn't indexed in any way, so this would be a linear search, which could be slow if you have a *lot* of users. 

If you want to map out what you think would be a reasonable error handling / conflict resolution strategy for this, that will help.

**Comment by iverok** [comment]
Ok, so we have looked a bit more into this now and we are looking at the following process:

1) A new table mapping Vipps id's (phone numbers essentiall) to users. This will be primarily used for login.
2) If the user is not mapped here, the (verified) email address is used, as it is today. If found, the user is mapped in the table mentioned above.
3) When "breaking the connection", we delete the table entry as well as the metadata stored for the user; thus allowing the user to log into another account for which the address is known.
4) We do not update the email address if the user is found via step 1.  This would presumably be because the user wants a different email address on this site. This also means that we will not log into any other accounts with the email address stored in the Vipps app in this case. It also means no new account will be created.

In short, "linkage" with existing users based on email will be done just once, at first login.

The rationale for this is that even though this does not handle all cases and is not maximally flexible, it will require a lot less code and maintenance, and no extra decision by the users (in the form of "which user are you trying to login as? should we update email addresses? etc). We will also not need to create any defaults for this.  For the normal case, this will work exactly as it does currently.

**Comment by iverok** [comment]
This is ready for next release.  For *connecting* to an account, there is still a requirement of same-email-addresses, but this can be filtered, as can the user lookup. 

The mapping table can in principle support many-to-many maps between Vipps accounts and WP accounts; but currently it's 1-1. This is because this does not require a user interface.  

It *is* possible to use the "confirmation screen" feature to remove these limitations, which would allow screens like "You are trying to connect with an account that has a different veriified email address. Are you sure",  "Which of your connected accounts are you trying to log in to" etc.  But this version does not do this. It does however allow changing email addresses, using phone numbers to find users etc.

**Comment by iverok** [comment]
Relased in 1.2.0
